block.go

```diff
@@ -270,6 +273,8 @@ func OpenBlock(dir string, pool chunkenc.Pool) (*Block, error) {
 		return nil, err
 	}
 
+	updateBlockSize(logger, dir, cr, meta)
```

Comment:
It would definitely work, I'm just not sure what the benefit is per se.

Reply:
Nope, I'm convinced. Just curious what the motivation was :)

Reply:
Should `updateBlockSize` still be a function that separates the blockSize call and errors from `OpenBlock`?

Reply:
sorry didn't understand?

why would we need the `updateBlockSize` ?

the meta write hapens in the else statement if this is what you mean.
```
else {
    meta.Stats.NumBytes = bs
    err := writeMetaFile(dir, meta)
    if err != nil {
		level.Warn(logger).Log("msg", "couldn't write the meta file for the block size", "block", dir, "err", err)
     }
}
```

Reply:
Also, one more thing. The meta, ir, and tr would all need to know the dir too, wouldn't they?

Reply:
I didn't know if you wanted to have this logic
```
bs,err:=blockSize( cr, ir, tr, meta) ; err != nil {
   level.Warn(logger).Log("msg", "couldn't get block size", "block", dir, "err", err)
} else {
    meta.Stats.NumBytes = bs
    err := writeMetaFile(dir, meta)
    if err != nil {
		level.Warn(logger).Log("msg", "couldn't write the meta file for the block size", "block", dir, "err", err)
     }
}
```

isolated from the OpenBlock method

Reply:
yeah. I think so

 for example:

 memTombstones should have a new field dir 
```
readTombstones(dir string) (*memTombstones, error) {
memTombstones.dir ==dir

func (t *memTombstones) Size() (int64,err){
   stat t.dir .....
}
...

Reply:
actually don't even need a new interface just add the Size function to 

```
// BlockReader provides reading access to a data block.
type BlockReader interface {
        // Size returns the block size in bytes.
        Size() int64,error
	// Index returns an IndexReader over the block's data.
	Index() (IndexReader, error)

	// Chunks returns a ChunkReader over the block's data.
	Chunks() (ChunkReader, error)

	// Tombstones returns a TombstoneReader over the block's deleted data.
	Tombstones() (TombstoneReader, error)
}
```

Reply:
> I didn't know if you wanted to have this logic

> bs,err:=blockSize( cr, ir, tr, meta) ; err != nil {
   level.Warn(logger).Log("msg", "couldn't get block size", "block", dir, "err", err)
} else {
    meta.Stats.NumBytes = bs
    err := writeMetaFile(dir, meta)
    if err != nil {
		level.Warn(logger).Log("msg", "couldn't write the meta file for the block size", "block", dir, "err", err)
     }
}
> ... isolated from the OpenBlock method


this is small enough to keep it there.

Reply:
Why wouldn't you need a new interface? ChunkReader, IndexReader, BlockMeta and TombstoneReader don't implement BlockReader, so adding Size() there wouldn't allow us to call size on cr, ir, tr, and meta.

Reply:
yes you are right , sorry.

Reply:
So for adding a `Size()` function to BlockMeta, it would need the dir to do a stat, which is stored in Block. It doesn't really seem to make sense to store `dir` in BlockMeta, but I'm not sure how to get that path to it without changing the `Size()` to be `Size(dir)`. Any suggestions?

Reply:
Yeah I just looked as well and it is a tricky one.
To be honest I don't think we should even include the meta files when calculating the size. These are only 200-300 bytes.

Reply:
For the tombstones, I'm running into another hiccup. I changed the signature of `NewMemTombstones()` to `NewMemTombstones(dir string)` so that I could stat the tombstones. But there's several references to `NewMemTombstones`, one is in `NewHead`, and in `LoopChunkSeries`. Neither of those functions has the data dir currently. 

I could modify `NewHead` to take in dir, because the only reference to `NewHead` comes from `block.Open` which has the dir. `LoopChunkSeries` is a bit weirder, it doesn't seem to have a great option for supplying the dir. One (not very good) option is to take dir from the `IndexReader` that now stores the index, which is in the `blockQuerier`. I don't really like this approach at all.

One other option is to change the Size() function to take in a dir. It doesn't seem logical for a size function to need to take in the dir, but it simplifies a lot of this and results in less changes in other places.

Any thoughts?

Reply:
I just had another look and `MemTombstones` doesn't even look like a good place for this. Since these are in memory struct so it is not a very logical to hold disk file size information.

This is what I came up:
```
// SizeReader .....
type SizeReader interface {
	Size() int
}

// TombstoneFile .....
type TombstoneFile struct {
	size int
}

// Size .....
func (t *TombstoneFile) Size() int {
	return t.size
}

func readTombstones(dir string) (*memTombstones, SizeReader, error) {
	b, err := ioutil.ReadFile(filepath.Join(dir, tombstoneFilename))
	if os.IsNotExist(err) {
		return NewMemTombstones(), nil, nil
	} else if err != nil {
		return nil, nil, err
	}

	sr := &TombstoneFile{
		size: len(b),
	}
```

Reply:
probably can use the same logic for the meta file in `readMetaFile`